### PR TITLE
Add a non-interactive configure mode

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
There is now a `--config` command which allows the user to specify the configuration keys as JSON. It also has some special handling:

1. `--config @-` will read the JSON from STDIN
2. `--config @filepath` will read the JSON from the given file path

A `--force` flag has also been added, which causes the configuration script to run even if the config hasn't changed.

Finally the `info` command has been re-implemented to display the configuration keys

### CODE REVIEW

File: lib/service/commands/info.rb [hunk 1]
```diff
@@ -27,16 +27,45 @@
...
+      <%= text %>
+
+      <% values.each do |opts| -%>
+      ## <%= opts['label'] %>
+      *Key:* <%= opts['key'] %>
+      *Default:* <%= opts['value'] %>
+      *Value:* <%= data[opts['key']] %>
+
+      <% end -%>
+      ERB
```
Comment (benarmston) [inline lib/service/commands/info.rb]: The output used here contains ruby warnings and is not easily scriptable.

```
[flight@gateway1 (ben1) ~]$ flight service info www 
/opt/flight/opt/service/vendor/ruby/2.7.0/gems/kramdown-1.16.2/lib/kramdown/converter/base.rb:102: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
/opt/flight/opt/service/vendor/ruby/2.7.0/gems/tty-markdown-0.6.0/lib/tty/markdown/parser.rb:15: warning: The called method `initialize' is defined here
Flight environment web server

Please provide values for the following configuration items.

  HTTP Port
  Key: port
  Default: 80
  Value:

  HTTPs Port
  Key: https_port
  Default: 443
  Value:
```

I think we'd be better of using our existing table output here.

File: lib/service/type.rb [hunk 2]
```diff
@@ -66,6 +66,19 @@ def all
...
+      # In these cases, the service should be skipped
+      def safe_enabled_types
+        @enabled_types ||= enabled.map do |id|
+          begin
+            self[id]
+          rescue UnknownServiceTypeError
+            $stderr.puts "Skipping unknown service: #{id}"
+          end
+        end.reject(&:nil?)
+      end
```
Comment (benarmston) [inline lib/service/type.rb]: Good catch with this issue.

The naming of this method and `enabled` could do with some work.  I was initially confused that this was returning an array of a different class of objects to `enabled`.  A better naming would be:

* Rename `Type.enabled` to `Type.enabled_names`
* Similar with `Type#enabled`
* Rename `Type.safe_enabled_types` to `Type.enabled`.

File: lib/service/commands/configure.rb [hunk 3]
```diff
@@ -67,6 +85,21 @@ def data
         @data ||= load
       end
 
+      def config_input
+        if options.config == '@-'
+          $stdin.read
```
Comment (benarmston) [inline lib/service/commands/configure.rb]: Can we support `/dev/stdin` as another spelling of STDIN please?

**Review by benarmston — CHANGES_REQUESTED** [review]
The non-interactive configuration looks mostly good.  I've added additional requests to its handling of stdin.

The output of the `info` command doesn't look right.  I think we'd be better off using our normal table formatting.  That way its consistent with most everywhere else.

File: lib/service/commands/info.rb (hunk 1 above)
Comment (WilliamMcCumstie) [reply lib/service/commands/info.rb]: https://github.com/openflighthpc/flight-service/pull/8/commits/d69e43548152a63a00ba1a597f9bd39e26df4193

File: lib/service/type.rb (hunk 2 above)
Comment (WilliamMcCumstie) [reply lib/service/type.rb]: I have taken a different approach because:

* `enabled_names` is a misnomer as missing `types` are not enabled (they are missing),
* It isn't immediately clearer if a `enabled` or `enabled_names` should be used, and
* I dislike how `enabled_names` opens the eigenclass to patch in a `save` method.

Instead the methods have been refactored into an `enabled` and `save_enabled_file` methods. This means:
* `Type.enabled` returns a regular array of enabled `Type` objects, and
* `Type.save_enabled_file` will save the current state of the previous method.

NOTE: This does mean "missing enabled types" will be implicitly disabled the next time the file is saved. This is because they do not appear in  the `enabled` list. This should not be an issue, as the missing `type` has probably been uninstalled.

Comment (WilliamMcCumstie) [reply lib/service/type.rb]: 50250a1dd66f49543eaf0d209f9690f46b0d3740

File: lib/service/commands/configure.rb (hunk 3 above)
Comment (WilliamMcCumstie) [reply lib/service/commands/configure.rb]: https://github.com/openflighthpc/flight-service/pull/8/commits/48a86a870bd0c629b17fb7bc254bd600b0f4d067

File: lib/service/type.rb (hunk 2 above)
Comment (benarmston) [reply lib/service/type.rb]: This has introduced a change in behaviour.  Consider the following:

```
sudo yum install flight-login-api
flight service enable login-api
sudo yum remove flight-login-api
flight service disable www; flight service enable www # one of these will have caused a change.
sudo yum install flight-login-api
```

Previously, following the reinstallation of `flight-login-api` it would have been enabled.  With this change, it is now disabled.

This is at odds with how `systemctl`/`systemd` does things.  There uninstalling services does not change their enabled status.

Perhaps, `save_enabled_file` could call `enabled(include_missing: true)`?

File: lib/service/commands/info.rb (hunk 1 above)
Comment (benarmston) [reply lib/service/commands/info.rb]: Looks much better.  Thanks.

File: lib/service/type.rb (hunk 2 above)
Comment (WilliamMcCumstie) [reply lib/service/type.rb]: I agree that this is a change in behaviour but I question if it matters because:
* The previous behaviour was to break if a type was missing,
* We don't have parity with `systemd` in multiple other ways, and
* Keeping them "pseudo-enabled" means they should be a valid input to `flight service disable`

Making them a valid input to `flight service disable` is probably more effort than it is worth as `Type['missing']` raises an error. It also means `Type#disable` can not be used to disable these "pseudo-enabled" types.

Something to consider before we add complexity to a rather esoteric use case.

Comment (benarmston) [reply lib/service/type.rb]: We seem to be talking at cross purposes here.  Let me try again.

I want the following interaction to *not* change.  That is, the enabled/disabled status is remembered on reinstall.

```
[flight@gateway1 (ben1) ~]$ flight service avail | grep console-api | cut -f 4
enabled
[flight@gateway1 (ben1) ~]$ sudo yum -y remove flight-console-api > /dev/null
[flight@gateway1 (ben1) ~]$ ( flight service enable www ; flight service disable www ) > /dev/null
[flight@gateway1 (ben1) ~]$ sudo yum -y install flight-console-api > /dev/null
[flight@gateway1 (ben1) ~]$ flight service avail | grep console-api | cut -f 4
enabled
```

I also want the following interaction to *not* change.  That is, attempting to enable or disable a uninstalled (or simply mis-typed) service results in an error.

```
[flight@gateway1 (ben1) ~]$ flight service enable console-api
Service 'console-api' already enabled.
[flight@gateway1 (ben1) ~]$ sudo yum -y remove flight-console-api > /dev/null
[flight@gateway1 (ben1) ~]$ flight service enable console-api
flight service: unknown service type: console-api
```

I want the following interaction to change.

```
[flight@gateway1 (ben1) ~]$ flight service stack status
┌──────────────────┬────────┬───────┐
│ Service          │ Status │ PID   │
├──────────────────┼────────┼───────┤
│ desktop-restapi  │ active │ 20947 │
│ file-manager-api │ active │ 16482 │
│ job-script-api   │ active │ 21140 │
│ login-api        │ active │ 16423 │
│ console-api      │ active │ 19534 │
└──────────────────┴────────┴───────┘
[flight@gateway1 (ben1) ~]$ sudo yum -y remove flight-console-api > /dev/null
[flight@gateway1 (ben1) ~]$ flight service stack status
flight service: unknown service type: console-api
```

The implementation as of 90ce9a023e83ea36e745c9467befc6da58122b53 gets the first of these wrong.  It gets the second correct and fixes the issue in the third.

Comment (WilliamMcCumstie) [reply lib/service/type.rb]: https://github.com/openflighthpc/flight-service/pull/8/commits/cae9a2b10b9cfc7df653ca9a08142e89f42e1f60

**Comment by benarmston** [comment]
After some consideration of whether the following would have been a better fix to the missing enabled types bug, I've decided to merge this as is.  The bellow is probably the better solution, but I've tested this implementation and don't want to make this change at this point.

```diff
diff --git a/lib/service/commands/stack.rb b/lib/service/commands/stack.rb
index 16a7c21..59a8130 100644
--- a/lib/service/commands/stack.rb
+++ b/lib/service/commands/stack.rb
@@ -33,7 +33,7 @@ module Service
   module Commands
     class Stack < Command
       def run
-        if Type.enabled.any?
+        if enabled_types.any?
           case action
           when 'start'
             puts "Starting enabled services:\n\n"
@@ -60,21 +60,21 @@ module Service
         if $stdout.tty?
           Table.emit do |t|
             headers 'Service', 'Status', 'PID'
-            Type.enabled.each do |t|
+            enabled_types.each do |t|
               row Paint[t.name, :cyan],
                   t.running? ? 'active' : 'stopped',
                   t.pid
             end
           end
         else
-          Type.enabled.each do |t|
+          enabled_types.each do |t|
             puts [t.name, t.running? ? 'active' : 'stopped', t.pid].join("\t")
           end
         end
       end
 
       def start
-        Type.enabled.each do |service|
+        enabled_types.each do |service|
           print "   > "
           if service.running?
             text = Paint["Service already running: #{service.name}", '#2794d8']
@@ -93,7 +93,7 @@ module Service
       end
 
       def restart
-        Type.enabled.each do |service|
+        enabled_types.each do |service|
           print "   > "
           status_text = Paint["Restarting service: #{service.name}", '#2794d8']
           begin
@@ -107,7 +107,7 @@ module Service
       end
 
       def stop
-        Type.enabled.each do |service|
+        enabled_types.each do |service|
           print "   > "
           if !service.running?
             text = Paint["Service already stopped: #{service.name}", '#2794d8']
@@ -126,7 +126,7 @@ module Service
       end
 
       def reload
-        Type.enabled.each do |service|
+        enabled_types.each do |service|
           print "   > "
           if !service.running?
             text = Paint["Service not running: #{service.name}", '#2794d8']
@@ -168,6 +168,17 @@ module Service
       ensure
         Whirly.stop
       end
+
+      def enabled_types
+        Types.enabled.map do |svc|
+          begin
+            Type[svc]
+          rescue UnknownServidiff --git a/lib/service/commands/stack.rb b/lib/service/commands/stack.rb
index 16a7c21..59a8130 100644
--- a/lib/service/commands/stack.rb
+++ b/lib/service/commands/stack.rb
@@ -33,7 +33,7 @@ module Service
   module Commands
     class Stack < Command
       def run
-        if Type.enabled.any?
+        if enabled_types.any?
           case action
           when 'start'
             puts "Starting enabled services:\n\n"
@@ -60,21 +60,21 @@ module Service
         if $stdout.tty?
           Table.emit do |t|
             headers 'Service', 'Status', 'PID'
-            Type.enabled.each do |t|
+            enabled_types.each do |t|
               row Paint[t.name, :cyan],
                   t.running? ? 'active' : 'stopped',
                   t.pid
             end
           end
         else
-          Type.enabled.each do |t|
+          enabled_types.each do |t|
             puts [t.name, t.running? ? 'active' : 'stopped', t.pid].join("\t")
           end
         end
       end
 
       def start
-        Type.enabled.each do |service|
+        enabled_types.each do |service|
           print "   > "
           if service.running?
             text = Paint["Service already running: #{service.name}", '#2794d8']
@@ -93,7 +93,7 @@ module Service
       end
 
       def restart
-        Type.enabled.each do |service|
+        enabled_types.each do |service|
           print "   > "
           status_text = Paint["Restarting service: #{service.name}", '#2794d8']
           begin
@@ -107,7 +107,7 @@ module Service
       end
 
       def stop
-        Type.enabled.each do |service|
+        enabled_types.each do |service|
           print "   > "
           if !service.running?
             text = Paint["Service already stopped: #{service.name}", '#2794d8']
@@ -126,7 +126,7 @@ module Service
       end
 
       def reload
-        Type.enabled.each do |service|
+        enabled_types.each do |service|
           print "   > "
           if !service.running?
             text = Paint["Service not running: #{service.name}", '#2794d8']
@@ -168,6 +168,17 @@ module Service
       ensure
         Whirly.stop
       end
+
+      def enabled_types
+        Types.enabled.map do |svc|
+          begin
+            Type[svc]
+          rescue UnknownServiceTypeError
+            nil
+          end
+        end
+          .compact
+      end
     end
   end
 end
ceTypeError
+            nil
+          end
+        end
+          .compact
+      end
     end
   end
 end
```
